src/mscorlib/src/System/Text/ASCIIEncoding.cs

```diff
@@ -78,12 +80,157 @@ public override unsafe int GetByteCount(char* chars, int count)
+                    }
+                }
+            } 
+            else 
+            {
+                // Nothing to encode
+                bytesWritten = 0;
+            }
+
+            return bytesWritten;
+        }
+
+        public override byte[] GetBytes(char[] chars) {
+            if (chars == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.chars, ExceptionResource.ArgumentNull_Array);
+            Contract.EndContractBlock();
+            
+            return GetBytesValidated(chars, 0, chars.Length);
+        }
+
+        public override byte[] GetBytes(char[] chars, int index, int count) {
+            if ((chars == null) ||
+                (index < 0) ||
+                (count < 0) ||
+                (chars.Length - index < count))
+            {
+                EncodingForwarder.ThrowValidationFailedException(chars, index, count);
+            }
+            Contract.EndContractBlock();
+            
+            return GetBytesValidated(chars, index, count);
+        }
+
+        private unsafe byte[] GetBytesValidated(char[] chars, int index, int count)
+        {
+            byte[] bytes;
+            if (count > 0)
+            {
+                fixed (char* input = chars)
+                {
+                    bytes = GetBytesValidated(input + index, count);
+                }
+            }
+            else
+            {
+                bytes = Array.Empty<byte>();
+            }
+
+            return bytes;
+        }
+
+        private unsafe byte[] GetBytesValidated(char* input, int charCount)
+        {
+            int remaining = 0;
+            // Assume string is all ASCII and size array for that
+            byte[] bytes = new byte[charCount];
+
+            int bytesWritten;
+            fixed (byte* output = &bytes[0]) 
+            {
+                int charactersConsumed;
+                if (!EncodingForwarder.TryEncode(input, charCount, output, charCount, out charactersConsumed, out bytesWritten)) 
+                {
+                    // Not all ASCII, get the byte count for the remaining encoded conversion
+                    remaining = GetByteCount(input + charactersConsumed, charCount - charactersConsumed, null);
+                }
+            }
+
+            if (remaining > 0) 
+            {
+                // Not all ASCII, fallback to slower path for remaining encoding
+                var encoded = ResizeGetRemainingBytes(input, charCount, ref bytes, bytesWritten, remaining);
+                Debug.Assert(encoded == remaining);
+            }
+
+            return bytes;
         }
 
+        private unsafe int ResizeGetRemainingBytes(char* chars, int charCount, ref byte[] bytes, int alreadyEncoded, int remaining)
+        {
+            // Resize the array to the correct size
+            byte[] oldArray = bytes;
+            bytes = new byte[alreadyEncoded + remaining];
```

Comment:
This changes the allocation profiles. It makes the change win some lose some, unclear whether it is net benefit for  programs out there. These allocating convenience methods are unlikely to be used on performance critical paths. It may be better to not touch them.